src/coreclr/System.Private.CoreLib/src/System/Enum.CoreCLR.cs

```diff
@@ -13,35 +13,73 @@ public abstract partial class Enum
         [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "Enum_GetValuesAndNames")]
         private static partial void GetEnumValuesAndNames(QCallTypeHandle enumType, ObjectHandleOnStack values, ObjectHandleOnStack names, Interop.BOOL getNames);
 
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        private static extern object InternalBoxEnum(RuntimeType enumType, long value);
+        private static unsafe object InternalBoxEnum(RuntimeType enumType, long value)
+        {
+            MethodTable* pMethodTable = (MethodTable*)enumType.m_handle;
+            object obj = GetEnumInfoForActivation(enumType).CreateUninitializedInstance(enumType);
+            ref byte dataRef = ref obj.GetRawData();
+
+            switch (pMethodTable->GetNumInstanceFieldBytes())
+            {
+                case 1:
+                    dataRef = (byte)value;
+                    break;
+                case 2:
+                    Unsafe.As<byte, ushort>(ref dataRef) = (ushort)value;
+                    break;
+                case 4:
+                    Unsafe.As<byte, uint>(ref dataRef) = (uint)value;
+                    break;
+                case 8:
+                    Unsafe.As<byte, ulong>(ref dataRef) = (ulong)value;
+                    break;
+                default:
+                    Debug.Fail("Unexpected enum size");
+                    break;
+            }
+
+            GC.KeepAlive(enumType);
+
+            return obj;
+        }
 
         [MethodImpl(MethodImplOptions.InternalCall)]
         private extern CorElementType InternalGetCorElementType();
 
         [MethodImpl(MethodImplOptions.InternalCall)]
         internal static extern RuntimeType InternalGetUnderlyingType(RuntimeType enumType);
 
-        private static EnumInfo GetEnumInfo(RuntimeType enumType, bool getNames = true)
+        /// <summary>
+        /// Gets an <see cref="EnumInfo"/> instance to use to create new boxed instances.
+        /// </summary>
+        /// <param name="enumType">The enum type to use.</param>
+        /// <remarks>
+        /// This is like <see cref="GetEnumInfo(RuntimeType, bool)"/>, but skips unnecessary initialization.
+        /// </remarks>
+        private static EnumInfo GetEnumInfoForActivation(RuntimeType enumType)
         {
-            EnumInfo? entry = enumType.GenericCache as EnumInfo;
+            if (enumType.GenericCache is not EnumInfo entry)
+            {
+                enumType.GenericCache = entry = new EnumInfo();
```

Comment:
Has rather less to do with this PR. However, I wonder why EnumInfo is a class and not a struct.
Does EnumInfo have to be a reference type?

Reply:
If it were a struct it would be boxed when storing into GenericCache.